src/nbl/video/IGPUDescriptorSet.cpp

```diff
@@ -5,35 +5,104 @@
         // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
-        std::uninitialized_default_construct_n(m_pool->getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
+        std::uninitialized_default_construct_n(m_pool->getDescriptorStorage(type) + m_storageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
     }
 
     const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
     if (mutableSamplerCount > 0)
-        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
+        std::uninitialized_default_construct_n(m_pool->getMutableSamplerStorage() + m_storageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)], mutableSamplerCount);
 }
 
 IGPUDescriptorSet::~IGPUDescriptorSet()
 {
     if (!isZombie())
     {
-        m_pool->deleteSetStorage(this);
-        if (m_pool->m_creationParameters.flags.hasFlags(IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT))
-            m_pool->m_descriptorSetAllocator.free_addr(m_poolOffset, 1);
+        auto dummy = this;
+        m_pool->deleteSetStorage(dummy);
     }
 }
 
+bool IGPUDescriptorSet::processWrite(const IGPUDescriptorSet::SWriteDescriptorSet& write)
+{
+    assert(write.dstSet == this);
+
+    system::ILogger* logger = nullptr;
+    {
+        auto debugCallback = getOriginDevice()->getPhysicalDevice()->getDebugCallback();
+        if (debugCallback)
+            logger = debugCallback->getLogger();
+    }
+
+    auto* descriptors = getDescriptors(write.descriptorType, write.binding);
+    if (!descriptors)
+    {
+        if (logger)
+            logger->log("Descriptor set layout doesn't allow descriptor of such type at binding %u.", system::ILogger::ELL_ERROR, write.binding);
+        return false;
+    }
+
+    core::smart_refctd_ptr<video::IGPUSampler>* mutableSamplers = nullptr;
+
+    if (write.descriptorType == asset::IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER && write.info->info.image.sampler)
+    {
+        mutableSamplers = getMutableSamplers(write.binding);
+        if (!mutableSamplers)
+        {
+            if (logger)
+                logger->log("Descriptor set layout doesn't allow mutable samplers at binding %u.", system::ILogger::ELL_ERROR, write.binding);
+            return false;
+        }
+    }
+
+    for (auto j = 0; j < write.count; ++j)
+    {
+        descriptors[j] = write.info[j].desc;
+
+        if (mutableSamplers)
+            mutableSamplers[j] = write.info[j].info.image.sampler;
+    }
+
+    incrementVersion();
+
+    return true;
+}
+
+bool IGPUDescriptorSet::processCopy(const IGPUDescriptorSet::SCopyDescriptorSet& copy)
+{
+#if 0
```

Comment:
whats with the ifdef?

Reply:
> also `incrementVersion` is missing

See https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1122668397